### PR TITLE
Add GBinding bindings

### PR DESCRIPTION
**Pull request by sdroege**
Kind of ugly because of the autogenerated code that can't be autogenerated due to how gir works :)

Also untested so far, please review carefully. Will report back once I know that it works.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Should we split apart GObject and GLib?

**Comment by sdroege** [comment]
> Should we split apart GObject and GLib?

GLib by itself is rather useless in Rust IMHO, and the code is all entangled (just because of the GType related things already, which are from GObject but used in GLib).

**Comment by GuillaumeGomez** [comment]
Indeed but it's a requirement that we need for the moment. And right now we can't generate two bindings at once if I understand things correctly.

**Comment by sdroege** [comment]
> Indeed but it's a requirement that we need for the moment. And right now we can't generate two bindings at once if I understand things correctly.

I would prefer to fix gir so it can output to e.g. `src/auto_gobject` via some commandline parameter

File: src/object.rs [hunk 1]
```diff
@@ -931,6 +934,57 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
...
+                )
+            )
+        }
+    }
+
+    fn bind_property_full<'a, 'b, O: IsA<Object>, N: Into<&'a str>, M: Into<&'a str>, F: Fn(&::Binding, &Value) -> Option<Value> + Send + Sync + 'static, G: Fn(&::Binding, &Value) -> Option<Value> + Send + Sync + 'static>(&self, source_property: N, target: &O, target_property: M, flags: ::BindingFlags, transform_to: F, transform_from: G) -> Option<::Binding> {
+        let source_property = source_property.into();
+        let target_property = target_property.into();
+
+        let to_closure = ::Closure::new(move |values| {
```
Comment (sdroege) [inline src/object.rs]: This is wrong, fixing

**Comment by EPashkin** [comment]
@sdroege I thought that path be `src/gobject/auto`, this way we can separate manual code too.
But your suggestion also acceptable.

**Comment by sdroege** [comment]
> @sdroege I thought that path be src/gobject/auto, this way we can separate manual code too.
> But your suggestion also acceptable.

It just has to be configurable, and more important gir has to create the use statements for GObject like for GLib

File: src/object.rs [hunk 2]
```diff
@@ -655,6 +655,9 @@ pub trait ObjectExt: IsA<Object> {
     ///
     /// [downgrade_sync]: trait.SyncObjectExt.html#method.downgrade_sync
     fn downgrade(&self) -> WeakRef<Self>;
+
+    fn bind_property<'a, 'b, O: IsA<Object>, N: Into<&'a str>, M: Into<&'a str>>(&self, source_property: N, target: &O, target_property: M, flags: ::BindingFlags) -> Option<::Binding>;
+    fn bind_property_full<'a, 'b, O: IsA<Object>, N: Into<&'a str>, M: Into<&'a str>, F: Fn(&::Binding, &Value) -> Option<Value> + Send + Sync + 'static, G: Fn(&::Binding, &Value) -> Option<Value> + Send + Sync + 'static>(&self, source_property: N, target: &O, target_property: M, flags: ::BindingFlags, transform_to: Option<F>, transform_from: Option<G>) -> Option<::Binding>;
```
Comment (sdroege) [inline src/object.rs]: This is suboptimal as you can't just pass `None`, it needs type annotations everywhere.

Opinions: This either needs 3 functions (first, second, both closures) or a builder (then all functions can be combined). IMHO a builder is nicer, then the flags could also become optional for example.

Comment (sdroege) [reply src/object.rs]: Builder it is if nobody complains :)

**Comment by sdroege** [comment]
Ok, please review the API if that makes sense to you :) One usage I have here right now is

```rust
        item.bind_property("age", &label, "label")
            .flags(glib::BindingFlags::DEFAULT | glib::BindingFlags::SYNC_CREATE)
            .transform_from(|_binding: &glib::Binding, value: &glib::Value| -> Option<glib::Value> {
                let age = value.get::<u32>().unwrap();
                Some(format!("{}", age).to_value())
            })
            .build();
```

File: src/object.rs [hunk 3]
```diff
@@ -1051,3 +1060,78 @@ impl<T: IsA<Object> + StaticType + UnsafeFrom<ObjectRef> + Wrapper + ?Sized> Wea
...
+    }
+
+    pub fn transform_to<F: Fn(&::Binding, &Value) -> Option<Value> + Send + Sync + 'static>(self, func: F) -> Self {
+        Self {
+            transform_to: Some(Self::transform_closure(func)),
+            ..self
+        }
+    }
+
+    pub fn flags(self, flags: ::BindingFlags) -> Self {
```
Comment (sdroege) [inline src/object.rs]: Instead of this, one could have a `fn bidirectional(self) -> Self` etc for each of the flags

**Comment by sdroege** [comment]
And autogenerated now

File: Gir_GObject.toml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ignore = [
+]
+
+manual = [
+    "GObject.Object",
+]
+
+[[object]]
+name = "GObject.*"
+status = "ignore"
```
Comment (EPashkin) [inline Gir_GObject.toml]: In final version this need be status = "generate"

Comment (sdroege) [reply Gir_GObject.toml]: Why? I don't care about any of those functions. They can be added when someone needs them :)

And more important, all that can be autogenerated need manual bindings (and many of them have already, either here or in my `gobject-subclass` crate)

File: src/object.rs [hunk 5]
```diff
@@ -1051,3 +1060,78 @@ impl<T: IsA<Object> + StaticType + UnsafeFrom<ObjectRef> + Wrapper + ?Sized> Wea
     }
 }
 
+pub struct BindingBuilder<'a, S: IsA<Object> + 'a, T: IsA<Object> + 'a> {
```
Comment (EPashkin) [inline src/object.rs]: Maybe this better moved to `src/gobject/binding_builder.rs `

File: Gir_GObject.toml (hunk 4 above)
Comment (EPashkin) [reply Gir_GObject.toml]: Ok

File: src/object.rs (hunk 5 above)
Comment (sdroege) [reply src/object.rs]: Ok, will move it later if @GuillaumeGomez has no other comments

File: src/gobject/auto/enums.rs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+// This file was generated by gir (https://github.com/gtk-rs/gir)
```
Comment (GuillaumeGomez) [inline src/gobject/auto/enums.rs]: Is it normal to have this empty file?

Comment (sdroege) [reply src/gobject/auto/enums.rs]: Yes

Comment (GuillaumeGomez) [reply src/gobject/auto/enums.rs]: Can we not have it if it's useless? :)

Comment (sdroege) [reply src/gobject/auto/enums.rs]: Sure, same for flags I guess :) But can we do that in a separate PR, it requires some further gir changes but I can put that on my list then

Comment (GuillaumeGomez) [reply src/gobject/auto/enums.rs]: Fine by me!

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**